### PR TITLE
Add Authenticator Pro encrypted import support, fixes #1035

### LINKED ISSUE #1035: Add support Import from Authenticator Pro
Hi, I really like Aegis
Pleaae add support Import from "Authenticator Pro" feature.
https://github.com/jamie-mh/AuthenticatorPro

### PR DESCRIPTION
**Pull request by NepNep21**
It also supports backups exported with the encrypted type but no password. 

Edit: Since [8db31d8](https://github.com/beemdevelopment/Aegis/pull/1080/commits/8db31d878bbb8a45acebee18cfb4b98b19243892) internal imports are also supported.

### CODE REVIEW

**Comment by ipkpjersi** [comment]
Since it's using encryption (not one-way hashing), would it not be possible to also support encrypted password backups too?

**Comment by NepNep21** [comment]
It is possible and already supported, if the raw file cannot be parsed as json, that is, it is encrypted, a password prompt appears which is used to decrypt it

**Comment by ipkpjersi** [comment]
Ohh right, I misunderstood what you meant by your first comment. Awesome!

**Comment by NepNep21** [comment]
Just noticed a bug that got past my original reread, a fixed commit should be ready soon (and more).

**Comment by NepNep21** [comment]
This commit fixes the mentioned bug, that being only 1 DatabaseImporterEntryException ever being reported due to jumping out of the loop. It also adds internal import support for rooted devices as a bonus.

**Comment by ipkpjersi** [comment]
Just tested it out, this seems to work well for me. I do have one suggestion, can we get it so that it doesn't set the name to "null"? 

Here's an example of what I'm talking about:


Hopefully a maintainer can take a look at reviewing this so it can hopefully get merged since it seems to work well.

**Comment by NepNep21** [comment]
Makes sense, I did not think of that, I'll look at it in a bit.

**Comment by NepNep21** [comment]
@ipkpjersi I was unable to find a test case where i got your result, however this change *should* hopefully fix it, could you try it?

**Comment by ipkpjersi** [comment]
Nope looks like they're still null, I made sure to clear all data then uninstall and clean install the latest version and verified the null check was in the code. Weird it still happens.

**Comment by NepNep21** [comment]
This should fix it now, I modified an export to create a test case and it works on my end.

**Comment by ipkpjersi** [comment]
Yep, looks like that did the trick, thanks.

**Comment by alexbakker** [comment]
Thanks for picking this up! Unfortunately I haven't had time to look at this yet, and I may not be able to get to it in the next couple of weeks either. Sorry about that. Starting February I should have some more time for Aegis.

File: app/src/main/java/com/beemdevelopment/aegis/importers/AuthenticatorProImporter.java [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+            this.algo = Algorithm.values()[SqlImporterHelper.getInt(cursor, "algorithm")];
+            this.digits = SqlImporterHelper.getInt(cursor, "digits");
+            this.period = SqlImporterHelper.getInt(cursor, "period");
+            this.counter = SqlImporterHelper.getInt(cursor, "counter");
+        }
+
+        @SuppressLint("DefaultLocale")
+        @NotNull
+        @Override
+        public String toString() {
```
Comment (alexbakker) [inline app/src/main/java/com/beemdevelopment/aegis/importers/AuthenticatorProImporter.java:262]: What is this used for?

**Review by alexbakker — CHANGES_REQUESTED** [review]
Apologies for the delay. I've added a couple of comments.

Couple of general remarks:
- By convention, our private fields are prefixed by ``_``. We do not refer to private fields through ``this`` but directly by their name.
- Could you add tests for this new importer? One for the SQLite format and one for the JSON format. The top of the ``DatabaseImporterTest`` class has some instructions for how to do so.

File: app/src/main/java/com/beemdevelopment/aegis/importers/AuthenticatorProImporter.java (hunk 1 above)
Comment (NepNep21) [reply app/src/main/java/com/beemdevelopment/aegis/importers/AuthenticatorProImporter.java:262]: It appears to be part of an old prototype that i didn't notice before committing. Will be removed.

Comment (NepNep21) [reply app/src/main/java/com/beemdevelopment/aegis/importers/AuthenticatorProImporter.java:262]: Ah  actually, it is used in https://github.com/beemdevelopment/Aegis/blob/8a22acc9959f031f2c672e53ad8234e2595d72b1/app/src/main/java/com/beemdevelopment/aegis/importers/AuthenticatorProImporter.java#L118
An idea i looked at other importers to see what they did regarding that

**Comment by NepNep21** [comment]
> Apologies for the delay. I've added a couple of comments.
> 
> Couple of general remarks:
> 
>     * By convention, our private fields are prefixed by `_`. We do not refer to private fields through `this` but directly by their name.
> 
>     * Could you add tests for this new importer? One for the SQLite format and one for the JSON format. The top of the `DatabaseImporterTest` class has some instructions for how to do so.

Almost done complying with the review, however there is one issue i can see, Authenticator Pro seems unable to import the last test qr code, presumably due to not supporting steam the same way Aegis does, https://github.com/jamie-mh/AuthenticatorPro/wiki/Importing-from-Steam

**Comment by alexbakker** [comment]
Changes look good, thanks.

> Authenticator Pro seems unable to import the last test qr code

That makes sense, we don't use ``otpauth://totp`` for Steam because it isn't actually TOTP. You can just enter that one into Authenticator Pro manually instead of scanning the QR code. Skipping the last one is also fine with me. The tests are really more about checking whether all importers still work if we make changes.

**Comment by NepNep21** [comment]
Alright, I will finish that when i get home in the evening.

**Comment by NepNep21** [comment]
Should be working now

**Comment by alexbakker** [comment]
Looks good. Could you squash your work into a single commit? Will merge this afterwards.

**Review by alexbakker — APPROVED** [review]

**Comment by NepNep21** [comment]
Done i think

**Comment by alexbakker** [comment]
Thanks!

**Pull request merged**